### PR TITLE
XOR games: nonsignaling value

### PR DESCRIPTION
**Pull request by juliusw352**
## Description
To address issue #15, I made a `nonsignaling_value` function that creates a `NonlocalGame` object and calls the already existing `nonsignaling_value` function.
(Edit: I have now also fixed a type indicator that I got wrong in last PR)

## Questions
-  For testing, i adapted the already existing `test_chsh_game_nonsignaling_value` and `test_chsh_game_nonsignaling_value_rep_2`. If there is a better way to do this, feel free to let me know.

## Status
-  [x] Ready to go

### CODE REVIEW

File: toqito/nonlocal_games/xor_game.py [hunk 1]
```diff
@@ -6,309 +6,316 @@
...
+		)
+	
+	def nonsignaling_value(self) -> float:
+		"""
+		Compute the nonsignaling value of an XOR game by first creating a more generic :code:`NonlocalGame` object and calling its :code:`nonsignaling_value` function.
+		
+		:return: A value between [0, 1] representing the nonsignaling value.
+		"""
+		nlg = self.to_nonlocal_game()
+		return nlg.nonsignaling_value()
```
Comment (vprusso) [inline toqito/nonlocal_games/xor_game.py:303]: We can make this a one-liner: 

```python
return self.to_nonlocal_game().nonsignaling_value()
```

File: toqito/nonlocal_games/xor_game.py [hunk 2]
```diff
@@ -6,309 +6,316 @@
...
+			return val
+		raise ValueError(
+			"Error: toqito currently does not support "
+			"multiple repetitions for the classical value of "
+			"a nonlocal game."
+		)
+	
+	def nonsignaling_value(self) -> float:
+		"""
+		Compute the nonsignaling value of an XOR game by first creating a more generic :code:`NonlocalGame` object and calling its :code:`nonsignaling_value` function.
```
Comment (vprusso) [inline toqito/nonlocal_games/xor_game.py:298]: This line is rather long. Perhaps we should keep the comments to 120 chars long.

File: tests/test_nonlocal_games/test_xor_game.py [hunk 3]
```diff
@@ -7,178 +7,199 @@
...
+		self.assertEqual(np.isclose(res, expected_res, atol=0.5), True)
+
+	def test_to_nonlocal_game(self):
+		"""XOR game to nonlocal game conversion."""
+		prob = 1/4 * np.ones((2,2))
+		pred = np.array([[0,0],[0,1]])
+
+		xor_chsh = XORGame(pred, prob, reps=1)
+		nlg_chsh = xor_chsh.to_nonlocal_game()
+
```
Comment (vprusso) [inline tests/test_nonlocal_games/test_xor_game.py:199]: Extra line here.

File: tests/test_nonlocal_games/test_xor_game.py [hunk 4]
```diff
@@ -7,178 +7,199 @@
...
+	def test_chsh_game_nonsignaling_value_rep_2(self):
+		"""Non-signaling value for the CHSH game for 2 reps."""
+		prob_mat = 1/4 * np.ones((2,2))
+		pred_mat = np.array([[0,0],[0,1]])
+
+		chsh = XORGame(prob_mat, pred_mat, 2)
+		res = chsh.nonsignaling_value()
+		expected_res = 1
+		self.assertEqual(np.isclose(res, expected_res, atol=0.5), True)
```
Comment (vprusso) [inline tests/test_nonlocal_games/test_xor_game.py:190]: This is a rather high `atol` value. I would suspect that we wouldn't need an `atol` param here at all, is that true?

File: tests/test_nonlocal_games/test_xor_game.py [hunk 5]
```diff
@@ -7,178 +7,199 @@
...
+
+	def test_chsh_game_nonsignaling_value(self):
+		"""Non-signaling value for the CHSH game."""
+		prob_mat = 1/4 * np.ones((2,2))
+		pred_mat = np.array([[0,0],[0,1]])
+
+		chsh = XORGame(prob_mat, pred_mat)
+		res = chsh.nonsignaling_value()
+		expected_res = 1
+		self.assertEqual(np.isclose(res, expected_res, atol=0.5), True)
```
Comment (vprusso) [inline tests/test_nonlocal_games/test_xor_game.py:180]: We can condense this down a bit:

```python
self.assertEqual(np.isclose(chsh.nonsignaling_value(), 1), True)
```

Could then remove the above variables. Same with the next test as well.

File: tests/test_nonlocal_games/test_xor_game.py [hunk 6]
```diff
@@ -7,178 +7,199 @@
...
+
+	def test_chsh_game_nonsignaling_value_rep_2(self):
+		"""Non-signaling value for the CHSH game for 2 reps."""
+		prob_mat = 1/4 * np.ones((2,2))
+		pred_mat = np.array([[0,0],[0,1]])
```
Comment (vprusso) [inline tests/test_nonlocal_games/test_xor_game.py:185]: As a nit-pick, PEP-8 suggests we put in spaces between comma-separated values. So for instance:

```python
np.array([[0,0],[0,1]])
```

would be: 

```python
np.array([[0, 0], [0, 1]])
```

etc.

File: toqito/nonlocal_games/xor_game.py [hunk 7]
```diff
@@ -6,309 +6,316 @@
...
+		"""
+		q_0, q_1 = self.prob_mat.shape
+		xor_pred_mat = self.pred_mat
+		nlg_pred_mat = np.ndarray((2,2,q_0,q_1))
+
+		for a in range(2):
+			for b in range(2):
+				for x in range(q_0):
+					for y in range(q_1):
+						nlg_pred_mat[a,b,x,y] = xor_pred_mat[x,y] == a ^ b
```
Comment (vprusso) [inline toqito/nonlocal_games/xor_game.py:319]: Same general comment about the comma-separated values.

File: tests/test_nonlocal_games/test_xor_game.py [hunk 8]
```diff
@@ -7,178 +7,199 @@
...
+
+		chsh = XORGame(prob_mat, pred_mat, 2)
```
Comment (vprusso) [inline tests/test_nonlocal_games/test_xor_game.py:187]: To make this a bit clearer, perhaps we want to slightly change this to:

```python
chsh_2_rep = XORGame(prob_mat, pred_mat, reps=2)
```

**Review by vprusso — COMMENTED** [review]
Excellent. Thank you again for the great contribution, @juliusw352 !

I added only a number of small comments, but overall looks great and it is an excellent usage of your `to_nonlocal_game` function.

As one final thing, I would suggest you add the following lines below the `#1.0.4` section in the `CHANGELOG.md`:

```
- Adding `to_nonlocal_game` function in `XORGame` that converts an XOR game
  definition to a `NonlocalGame` object (thanks to @juliusw352)

- Adding `nonsignaling` function in `XORGame` that calculates the nonsignaling
  value of an `XORGame` object (thanks to @juliusw352)
```

Of course, feel free to tweak as you see fit!

**Comment by juliusw352** [comment]
All done!

I also have a working `classical_value` function that supports parallel repetition, but it runs quite slow and looks rather messy – would it be a good idea to do something similar there as is done here by referencing the function in the `NonlocalGame` class if reps > 1?

**Comment by vprusso** [comment]
Awesome, thank you for the updates, @juliusw352 !

The changes look good, so I'll go ahead and merge. Thank you again for your contribution to `toqito`!

> I also have a working classical_value function that supports parallel repetition, but it runs quite slow and looks rather messy – would it be a good idea to do something similar there as is done here by referencing the function in the NonlocalGame class if reps > 1?

For this, I would suggest opening up a separate PR and we can see if it's possible to clean that up and optimize it. Thank you again!

**Pull request merged**